geomstats/numerics/geodesic.py

```diff
@@ -443,15 +404,12 @@ def log(self, space, point, base_point):
 
         res = self.optimizer.minimize(objective, init_tangent_vec)
 
-        tangent_vec = gs.reshape(res.x, base_point.shape)
-
-        return tangent_vec
+        return gs.reshape(res.x, base_point.shape)
 
 
-class _LogShootingSolverUnflatten(
-    _LogBatchMixins, _GeodesicBVPFromExpMixins, LogSolver
-):
+class _LogShootingSolverUnflatten(_LogBatchMixins, LogSolver):
```

Comment:
Docstring? At this point, I'm confused about why there are so many log classes.

Reply:
We have an algorithm: log shooting. From a user perspective we have one solver `LogShootingSolver`. We have two implementations of this algorithm that change only on the way they handle vectorization. The choice of the version from a user perspective is done with `flatten`:

>     flatten : bool
>        If True, the optimization problem is solved together for all the points.


If `True`, only one optimization problem is solved each time, even in batch mode. This is implemented by `_LogShootingSolverFlatten`. The name comes from the fact that the input is `flattened` before the optimization.

If `False`, an optimization problem is solved for each point/base_point in the batch.

The implementations are sufficiently different to deserve different classes (NB: this detail is hidden from the users.)

The flatten implementation (original implementation in geomstats) is less robust for the cases where the difficulty of each log problem in the batch is very different (number of iterations always depend on the worst case). It may be faster for similarly difficult logs (depends on the space).